tensorflow_quantum/python/differentiators/differentiator.py

```diff
@@ -179,6 +179,73 @@ def refresh(self):
+        different parameter derivatives.
+
+        Args:
+            programs: `tf.Tensor` of strings with shape [batch_size] containing
+                the string representations of the circuits to be executed.
+            symbol_names: `tf.Tensor` of strings with shape [n_params], which
+                is used to specify the order in which the values in
+                `symbol_values` should be placed inside of the circuits in
+                `programs`.
+            symbol_values: `tf.Tensor` of real numbers with shape
+                [batch_size, n_params] specifying parameter values to resolve
+                into the circuits specified by programs, following the ordering
+                dictated by `symbol_names`.
+            pauli_sums: `tf.Tensor` of strings with shape [batch_size, n_ops]
+                containing the string representation of the operators that will
+                be used on all of the circuits in the expectation calculations.
+
+        Returns:
+            flat_programs: 1-D `tf.Tensor` of strings representing all the
+                circuits required to be run to evaluate the gradient under the
+                derived differentiator.  The length is determined by the
+                specifics of the derived differentiator.
+            flat_symbol_names: 1-D `tf.Tensor` of of strings, containing the
+                name of every symbol used in every circuit in `flat_programs`.
+                The length is determined by the specifics of the derived
+                differentiator.
+            flat_symbol_values: 2-D `tf.Tensor` of DType `tf.float32` containing
+                values to fill in to every parameter in every circuit.
+                The first dimension is the length of `flat_programs` and the
+                second is the length of `flat_symbol_names`.
+            flat_pauli_sums: 2-D `tf.Tensor` of strings representing the
+                operators to measure against each circuit of `flat_programs`.
+                First dimension is the length of `flat_program`, but the
+                second dimension is determined by the specifics of the
+                derived differentiator.
+            flat_mapper: 5-D `tf.Tensor` of DType `tf.float32`.
+                The first dimension is the length of the input `programs`, the
+                second dimension is the length of the second dimension of the
+                input `pauli_sum`, the third dimension is the length of the
+                input `symbol_names`, and the last two dimensions are the same
+                shape as the output `flat_pauli_sums`.  The value at index
+                `ijkmn` is the amount of weight to give the measurement result
+                at `mn` when summing over expectation values to obtain the
+                gradient of the symbol `k` with respect to pauli sum `j` of
+                program `i`.
+        """
```

Comment:
Given the potentially drastically different "intermediate logic blocks" between differentiators. I think it might make more sense to just have the return signatures defined on a per differentiator basis and leave the parent class as an abstract method. Trying to fit all differentiators under this roof might prove pretty hard.

Reply:
Even though every differentiator can be defined via this interface, I understand your point that it could be a bit hard to reason about.  I think it could make sense to restrict the interface slightly to assume we will use the same gradient procedure for every circuit; then the mapper tensor would be the same for every circuit in the input batch, allowing us to drop an index and reduce the size of the mapper.  I'll think a bit more about how to simplify this interface along that line.

Reply:
Changed to a new, simpler batch interface.  ParameterShift and LinearCombination have been updated to obey the interface.  I still need to write tests for `get_intermediate_logic` confirming shapes and values explicitly.

Reply:
Now added an explicit equation at the end of the docstring to justify why the returns have the shapes specified.